File: binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> output;

    vector<int> rightSideView(TreeNode* root){

        helper(root, 0);
        return this->output;
    }

    int helper(TreeNode* root, int level){
        if (!root){
            return 0;
        }
        if (level == output.size()){
            this->output.push_back(root->val);
        }
        helper(root->right, level + 1);
        helper(root->left, level + 1);
    }



    vector<int> rightSideViewIterative(TreeNode* root) {
        vector<int> output;
        vector<tuple<TreeNode*, int> > nodes_levels;

        int levelmax = 0;
        int level = 0;
        TreeNode *node;

        nodes_levels.push_back(make_tuple(root, level));

        while (!nodes_levels.empty()){
            tie(node, level) = nodes_levels.back();
            nodes_levels.pop_back();
            while (node){
                if (node->left){
                    nodes_levels.push_back(make_tuple(node->left, level+1));
                }
                if (level == levelmax){
                    output.push_back(node->val);
                    levelmax += 1;
                }
                level += 1;
                node = node->right;
            }
        }
        return output;
    }
};

File: binary-tree-right-side-view.py
__author__ = 'stephenosullivan'

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """

        output = []
        level = 1
        levelmax = 0

        # Store nodes as a tuple with a level indicator
        nodes_to_visit = [(root, level)]

        # Sweep through the tree; 'right' branches first
        while nodes_to_visit:
            node, level = nodes_to_visit.pop()
            while node:
                if node.left:
                    nodes_to_visit.append((node.left, level + 1))
                if level > levelmax:
                    output.append(node.val)
                    levelmax += 1
                level += 1
                node = node.right

        return output
